cpp/src/arrow/sparse_tensor.cc

```diff
@@ -422,6 +424,139 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
   inline Status CheckMaximumValue(const uint64_t) const { return Status::OK(); }
 };
 
+// ----------------------------------------------------------------------
+// SparseTensorConverter for SparseCSFIndex
+
+template <typename TYPE>
+class SparseTensorConverter<TYPE, SparseCSFIndex>
+    : private SparseTensorConverterBase<TYPE> {
+ public:
+  using BaseClass = SparseTensorConverterBase<TYPE>;
+  using typename BaseClass::NumericTensorType;
+  using typename BaseClass::value_type;
+
+  SparseTensorConverter(const NumericTensorType& tensor,
+                        const std::shared_ptr<DataType>& index_value_type,
+                        MemoryPool* pool)
+      : BaseClass(tensor, index_value_type, pool) {}
+
+  template <typename IndexValueType>
+  Status Convert() {
+    using c_index_value_type = typename IndexValueType::c_type;
+    RETURN_NOT_OK(CheckMaximumValue(std::numeric_limits<c_index_value_type>::max()));
+
+    const int64_t ndim = tensor_.ndim();
+    std::vector<int64_t> axis_order = internal::ArgSort(tensor_.shape());
+
+    if (ndim < 2) {
+      // LCOV_EXCL_START: The following invalid causes program failure.
+      return Status::Invalid("Invalid tensor dimension");
+      // LCOV_EXCL_STOP
+    }
+
+    std::shared_ptr<SparseCOOTensor> sparse_coo_tensor;
+    ARROW_ASSIGN_OR_RAISE(sparse_coo_tensor, SparseCOOTensor::Make(tensor_));
+    std::shared_ptr<Tensor> coords =
+        arrow::internal::checked_pointer_cast<SparseCOOIndex>(
+            sparse_coo_tensor->sparse_index())
+            ->indices();
+
+    // TODO(rok): Coords should be sorted with axis_order priority to improve compression.
+    // ARROW-4221 would help here as well.
```

Comment:
@rok Why do you need ARROW-4221 (coords' canonical flag) to sort the coords tensor?
In the current implementation, the coords tensor in SparseCOOTensor should always be sorted in the index's lexicographical order.
Doesn't this restriction help you to sort the coords with axis_order?

Reply:
Changing axis order changes lexicographical order of the COO indices matrix. I've changed the approach now, so this is no longer an issue.